From: Andrey Rakhmatullin <wrar@debian.org>
Date: Fri, 17 Jan 2025 23:04:12 +0500
Subject: disable test_set requiring network access

---
 tests/test_ssl.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tests/test_ssl.py b/tests/test_ssl.py
index bcad6d9..f67432e 100644
--- a/tests/test_ssl.py
+++ b/tests/test_ssl.py
@@ -1309,6 +1309,7 @@ class TestContext:
         # internet which has such a certificate.  Connecting to the network
         # in a unit test is bad, but it's the only way I can think of to
         # really test this. -exarkun
+        return
         context = Context(SSLv23_METHOD)
         context.set_default_verify_paths()
         context.set_verify(
