diff --git a/OpenSSL/SSL.py b/OpenSSL/SSL.py
index a257f16..cfb0d5b 100644
--- a/OpenSSL/SSL.py
+++ b/OpenSSL/SSL.py
@@ -859,9 +859,10 @@ class Connection(object):
                         errno = _ffi.getwinerror()[0]
                     else:
                         errno = _ffi.errno
-                    raise SysCallError(errno, errorcode[errno])
-                else:
-                    raise SysCallError(-1, "Unexpected EOF")
+
+                    if errno != 0:
+                        raise SysCallError(errno, errorcode[errno])
+                raise SysCallError(-1, "Unexpected EOF")
             else:
                 # TODO: This is untested.
                 _raise_current_error()
