--- pyopenssl.orig/src/OpenSSL/SSL.py
+++ pyopenssl/src/OpenSSL/SSL.py
@@ -1924,6 +1924,33 @@ class Context:
             self._cookie_verify_helper.callback,
         )
 
+    def set_tmp_dh_from_pem(self, pem):
+        """
+        Set DH parameters from PEM-encoded data
+
+        :param pem: String containing PEM-encoded DH parameters.
+        :return: None
+        """
+        if not isinstance(pem, bytes):
+            raise TypeError("pem must be a byte string")
+
+        bio = _lib.BIO_new(_lib.BIO_s_mem())
+        if bio == _ffi.NULL:
+            _raise_current_error()
+        bio = _ffi.gc(bio, _lib.BIO_free)
+
+        result = _lib.BIO_write(bio, pem, len(pem))
+        if result <= 0:
+            _raise_current_error()
+
+        pkey = _lib.PEM_read_bio_Parameters(bio, _ffi.NULL)
+        _openssl_assert(pkey != _ffi.NULL)
+        res = _lib.SSL_CTX_set0_tmp_dh_pkey(self._context, pkey)
+        if res != 1:
+            _lib.EVP_PKEY_free(pkey)
+        _openssl_assert(res == 1)
+
+
 
 class Connection:
     _reverse_mapping: typing.MutableMapping[Any, Connection] = (
